Docker compose update (#2789)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/download-and-install/docker.mdx

All of the FusionAuth Docker images may be found on [Docker Hub](https://hub.docker.com/u/fusionauth/).

> The FusionAuth Docker Compose yaml files may be found on [FusionAuth containers repository](https://github.com/FusionAuth/fusionauth-containers) and more examples in the [FusionAuth docker compose example repository](https://github.com/sonderformat-llc/fusionauth-example-docker-compose) described in more detail below.

Reviewer: You missed this reference to sonderformat-llc.

The Docker Compose file references the `latest` tag, but that tag is not dynamic. It is only the latest at a point in time. To get the most recently released image, you have a couple of options.

* Pull the latest image with this command: `docker pull fusionauth/fusionauth-app:latest` or `docker compose pull` and recreate your deployment with `docker compose up -d` which will recreate every container where a new image is available.
* Edit the Docker Compose file and specify a specific version. This is a good idea for a production deployment. With `docker compose pull` you can only pull the specified image, or just run `docker compose up -d` which will pull the image and recreate the container at the same time.

> * Remove all images with `docker rmi <IMAGE ID>` selected from the list of IMAGE IDs which you can show with `docker images`. This requires that the image isn't used and therefore may prompt you to remove containers using it. Since all state of FusionAuth is stored in the database, and for docker you use volumes to persist data, you can safely remove the containers. After that you can use `docker compose pull`, `docker compose build` and `docker compose up -d` accordingly to get images specified the in the `docker-compose.yml`.

Reviewer: I've never needed to use `docker compose build`. Does `docker compose up -d` do that implicitly?

Author: There are some scenarios where you use `docker compose build` especially in combination with `--pull` and `--no-cache` in scenarios where you use the latest tag. Or when you want to first test the build on its own after an version update.

Reviewer: Okay, that's fine, let's leave it in there then.

> To generally clean up your images on your system `docker image prune` is useful too as it will remove all dangling images.

Reviewer: ```suggestion
To generally clean up your images on your system ,`docker image prune` will remove all unused, dangling images.
```

The Docker Compose file references the `latest` tag, but that tag is not dynamic. It is only the latest at a point in time. To get the most recently released image, you have a couple of options.

* Edit the Docker Compose file and specify a specific version. This is a good idea for a production deployment.

> * Remove the old image first: `docker rmi <old image id>`. This command may prompt you to remove containers using that image. Since all state is stored in the database, you can safely remove the containers. Then, when you recreate Docker Compose, the latest image will be retrieved.

Reviewer: You aren't recreating docker compose, you are restarting docker compose which recreates the network, pulls images, etc. What am I missing?

Author: I completely redone the section, and describing any of it as restart doesn't make sense in general, unless we use docker compose stop/start/restart commands.

* In the `volumes:` section of the `fusionauth` service, add `- ./kickstart:/usr/local/fusionauth/kickstart`.
* Modify `.env` and add the Kickstart configuration variable: `FUSIONAUTH_APP_KICKSTART_FILE=/usr/local/fusionauth/kickstart/kickstart.json`. This path should be what the Docker container expects, not the path on the host.
* Configure `docker-compose.yml` to pass the environment variable set by `.env` to the container. Do this by adding `FUSIONAUTH_APP_KICKSTART_FILE: ${FUSIONAUTH_APP_KICKSTART_FILE}` to the `environment` section of the `fusionauth` service. 
* `docker compose up -d`

The following is an example `docker-compose.yml` file configuring FusionAuth to run the commands in a `kickstart.json` at startup.

<RemoteCode title="Example docker-compose.yml for running Kickstart" url="https://raw.githubusercontent.com/sonderformat-llc/fusionauth-example-docker-compose/main/kickstart/docker-compose.yml" lang="yaml" />

> After running `docker compose up -d` you should see a line like this in the logs `docker compose logs fusionauth`:

Reviewer: ```suggestion
After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can  be accessed using this command: `docker compose logs fusionauth`:
```

* In the `volumes:` section of the `fusionauth` service, add `- ./kickstart:/usr/local/fusionauth/kickstart`.
* Modify `.env` and add the Kickstart configuration variable: `FUSIONAUTH_APP_KICKSTART_FILE=/usr/local/fusionauth/kickstart/kickstart.json`. This path should be what the Docker container expects, not the path on the host.
* Configure `docker-compose.yml` to pass the environment variable set by `.env` to the container. Do this by adding `FUSIONAUTH_APP_KICKSTART_FILE: ${FUSIONAUTH_APP_KICKSTART_FILE}` to the `environment` section of the `fusionauth` service. 
* `docker compose up -d`

The following is an example `docker-compose.yml` file configuring FusionAuth to run the commands in a `kickstart.json` at startup.

<RemoteCode title="Example docker-compose.yml for running Kickstart" url="https://raw.githubusercontent.com/FusionAuth/fusionauth-example-docker-compose/main/kickstart/docker-compose.yml" lang="yaml" />

> After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can  be accessed using this command: `docker compose logs fusionauth`:

Reviewer: ```suggestion
After running `docker compose up -d` you should see a line similar to the one below in the logs. These logs can be accessed using this command: `docker compose logs fusionauth`:
```

## Plugins

Instead of building a custom docker image, you can directly mount a directory containing a plugin to your Docker container.

Here are the steps to do so:

* In the `volumes:` section of the `fusionauth` service, add `- ./plugins:/usr/local/fusionauth/plugins`.
* Copy your plugin jar file, created [by following the instructions](/docs/extend/code/password-hashes/), to your `plugins` directory on the host.
* `docker compose up -d`

The following is an example `docker-compose.yml` file configuring FusionAuth to scan for plugins at startup.

<RemoteCode title="Example docker-compose.yml for installing a plugin" url="https://raw.githubusercontent.com/FusionAuth/fusionauth-example-docker-compose/main/plugin/docker-compose.yml" lang="yaml" />

> After running `docker compose up -d` you should see a line like this in the logs `docker compose logs fusionauth`:

Reviewer: ```suggestion
After running `docker compose up -d` you should see a line like this in the logs, which you can access using the command `docker compose logs fusionauth`:
```

## Accessing the Host Machine

The default FusionAuth Docker configuration sets up the network using the `bridge` configuration. This means that all the hosts defined by `docker-compose.yml` can access each other. However, it means that any applications running on your host machine cannot be accessed by FusionAuth using `localhost`.

This is typically only an issue when FusionAuth is accessing resources outside of the Docker network to, for example, send email or request a webhook. If a Rails application is running locally on a Mac and you want FusionAuth, running in Docker, to send a webhook payload to it, `localhost` won't work.

In this situation, do one of the following:

> * Use `host.docker.internal` as the hostname instead of `localhost`; this works on macOS and Windows hosts, for linux you have to add `--add-host=host.docker.internal:host-gateway` to the docker engine.

Reviewer: ```suggestion
* For macOS and Windows hosts, you can use `host.docker.internal` as the hostname instead of `localhost`.
* For Linux hosts, add `--add-host=host.docker.internal:host-gateway` to the docker engine.
```

Updated this for clarity, see next comment for correctness.

Reviewer: Can you explain what you mean by 'docker engine' here? Also, this post indicates you can do the same with the `extra_hosts` YAML key: https://stackoverflow.com/questions/70725881/what-is-the-equivalent-of-add-host-host-docker-internalhost-gateway-in-a-comp  

That seems preferable.

* Run a container with your application in Docker
* Install FusionAuth on the host machine, using FastPath or another method

Modifying the FusionAuth service in `docker-compose.yml` to use other Docker networking schemes such as `host` may work, but isn't fully tested or supported.

## OpenSearch Production Deployment Configuration

> The productive runtime requirements and configuration for OpenSearch will drastically differ from the docker compose examples as they're configured without any security and redundancy.

Reviewer: ```suggestion
Production runtime requirements and configuration for OpenSearch will drastically differ from the docker compose examples as the examples are configured without any security or redundancy.
```
